### PR TITLE
syncer: refine syncer test to trigger casuality

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
issue: https://github.com/pingcap/dm/issues/70
detect casuality conflict and generate flush job test case

### What is changed and how it works?
refine test, run some sqls which will trigger casuality 

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: syncer/syncer_test.go [hunk 1]
```diff
@@ -1203,26 +1235,28 @@ func (s *testSyncerSuite) TestRun(c *C) {
...
+
+		c.Assert(job.tp, Equals, expectJobs2[i].tp)
+		if job.tp == ddl {
+			c.Assert(job.ddls[0], Equals, expectJobs2[i].sqlInJob)
 		} else {
-			c.Assert(testJobs.jobs[i].sql, Equals, testCase.sqlInJob)
-			c.Assert(testJobs.jobs[i].args[0], Equals, testCase.arg)
+			c.Assert(job.sql, Equals, expectJobs2[i].sqlInJob)
+			c.Assert(job.args, DeepEquals, expectJobs2[i].args)
 		}
```
Comment (suzaku) [inline syncer/syncer_test.go]: It seems the execution of SQLs and the validation afterwards are almost identical for sql1 and sql2. Maybe we can extract a function or two to make the test simpler.

Comment (WangXiangUSTC) [reply syncer/syncer_test.go]: updated https://github.com/pingcap/dm/pull/198/commits/47740ebba6850394b74fddb1558349c147e359b8

Comment (suzaku) [reply syncer/syncer_test.go]: Can we do the same for the execution and wait part?

File: syncer/syncer_test.go [hunk 2]
```diff
@@ -1074,53 +1084,63 @@ func (s *testSyncerSuite) TestRun(c *C) {
...
-	testCases1 := []struct {
-		sql      string
-		tp       opType
-		sqlInJob string
-		arg      interface{}
-	}{
+	sqls1 := []string{
+		"create database if not exists test_1",
+		"create table if not exists test_1.t_1(id int primary key, name varchar(24))",
+		"create table if not exists test_1.t_2(id int primary key, name varchar(24))",
```
Comment (GregoryIan) [inline syncer/syncer_test.go]: what does need another table  to test ?

File: syncer/syncer_test.go [hunk 3]
```diff
@@ -1074,53 +1084,63 @@ func (s *testSyncerSuite) TestRun(c *C) {
...
 			"DELETE FROM `test_1`.`t_1` WHERE `id` = ? LIMIT 1;",
-			int64(580981944116838401),
+			[]interface{}{int64(580981944116838401)},
+		}, {
+			flush,
+			"",
+			nil,
+		}, {
+			update,
+			"DELETE FROM `test_1`.`t_1` WHERE `id` = ? LIMIT 1;",
```
Comment (GregoryIan) [inline syncer/syncer_test.go]: safe mode?

**Comment by GregoryIan** [comment]
Rest LGTM

File: syncer/syncer_test.go (hunk 3 above)
Comment (WangXiangUSTC) [reply syncer/syncer_test.go]: yes, I add a comment `in first 5 minutes, safe mode is true, will split update to delete + replace`

File: syncer/syncer_test.go (hunk 1 above)
Comment (WangXiangUSTC) [reply syncer/syncer_test.go]: add a new function `executeSQLAndWait`

File: syncer/syncer_test.go (hunk 2 above)
Comment (WangXiangUSTC) [reply syncer/syncer_test.go]: because I add `RouteRules` to route `t_1` to `t_2` when update config, so create table `t_2` here

**Comment by WangXiangUSTC** [comment]
@GregoryIan @suzaku PTAL again

**Pull request merged**